Running integration tests with lxc fail

So the `exec` tests should obviously fail, some of the `cap-add` ones make sense, and the networking ipc and mac address ones should fail as well, but there are a few additional tests that are failing that I don't think should fail.
Maybe @dineshs-altiscale can clear up some of the uncertainties here.
Failing tests:
```
--- FAIL: TestPsListContainersSize (1.22s)
	docker_cli_ps_test.go:281: Expected size "2 B", got "80 B"
--- FAIL: TestPsListContainersFilterStatus (1.98s)
	docker_cli_ps_test.go:322: Expected id 0680995f62db, got 0680995f62db
		cdede23854f1 for exited filter, output: "0680995f62db\ncdede23854f1\n"

--- FAIL: TestRunVolumesMountedAsReadonly (0.95s)
	docker_cli_run_test.go:368: run should fail because volume is ro: exit code 0
--- FAIL: TestRunVolumesFromInReadonlyMode (2.08s)
	docker_cli_run_test.go:384: run should fail because volume is ro: exit code 0
--- FAIL: TestRunVolumesFromInReadWriteMode (0.03s)
	docker_cli_run_test.go:396: exit status 1
--- FAIL: TestVolumesFromGetsProperMode (0.02s)
	docker_cli_run_test.go:422: exit status 1
--- FAIL: TestRunApplyVolumesFromBeforeVolumes (0.02s)
	docker_cli_run_test.go:449: exit status 1

--- FAIL: TestRunCreateVolumeWithSymlink (1.77s)
	docker_cli_run_test.go:534: [run] err: exit status 255, exitcode: 255
--- FAIL: TestRunVolumesFromSymlinkPath (2.13s)
	docker_cli_run_test.go:578: [run] (volume) err: exit status 255, exitcode: 255

--- FAIL: TestRunEnvironment (1.13s)
	docker_cli_run_test.go:762: Wrong environment: should be 8 variables, not: ", FALSE=true, HOME=/root, HOSTNAME=testing, PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin, TRICKY=tri, TRUE=false, cky, container=lxc"

--- FAIL: TestRunCapDropInvalid (1.03s)
	docker_cli_run_test.go:898: <nil> bin
		dev
		etc
		home
		lib
		lib64
		linuxrc
		media
		mnt
		opt
		proc
		root
		run
		sbin
		sys
		tmp
		usr
		var
		

--- FAIL: TestRunCapDropALLCannotMknod (1.00s)
	docker_cli_run_test.go:938: <nil> ok
		
--- FAIL: TestRunCapDropALLAddMknodCannotMknod (1.25s)
	docker_cli_run_test.go:953: exit status 1 2014/12/31 19:58:28 setup user setgroups operation not permitted
		lxc-start: The container failed to start.
		lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
		
--- FAIL: TestRunCapAddInvalid (1.28s)
	docker_cli_run_test.go:970: <nil> bin
		dev
		etc
		home
		lib
		lib64
		linuxrc
		media
		mnt
		opt
		proc
		root
		run
		sbin
		sys
		tmp
		usr
		var

--- FAIL: TestRunCapAddALLCanDownInterface (0.93s)
	docker_cli_run_test.go:995: exit status 255 lxc-start: unknown capability block_suspend
		lxc-start: failed to keep requested caps
		lxc-start: failed to setup the container
		lxc-start: invalid sequence number 1. expected 2
		lxc-start: failed to spawn 'f29df582a245c01a2e5a7dfa558f0647a7ac936e0f560e774b2440b9228df944'
		lxc-start: The container failed to start.
		lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
		

--- FAIL: TestRunSysNotWritableInNonPrivilegedContainers (1.04s)
	docker_cli_run_test.go:1056: sys should not be writable in a non privileged container
[PASSED]: run - sys writable in privileged container
--- FAIL: TestRunProcNotWritableInNonPrivilegedContainers (0.96s)
	docker_cli_run_test.go:1078: proc should not be writable in a non privileged container

--- FAIL: TestRunWriteHostsFileAndNotCommit (1.12s)
	docker_cli_run_test.go:1706: diff should be empty
--- FAIL: TestRunWriteHostnameFileAndNotCommit (1.16s)
	docker_cli_run_test.go:1735: diff should be empty
--- FAIL: TestRunWriteResolvFileAndNotCommit (1.07s)
	docker_cli_run_test.go:1764: diff should be empty

--- FAIL: TestRunUnknownCommand (1.24s)
	docker_cli_run_test.go:2480: Container should not have been able to start!

--- FAIL: TestSaveDirectoryPermissions (2.67s)
	docker_cli_save_load_test.go:425: failed to find the layer with the right content listinghttps://jenkins.dockerproject.com/job/Docker%20Master%20(iffy%20builds)/label=debian-devmapper-lxc/7/console
```

Full log can be seen here: https://jenkins.dockerproject.com/job/Docker%20Master%20(iffy%20builds)/label=debian-devmapper-lxc/7/console

```
--- FAIL: TestEventsContainerFailStartDie (1.33s)
	docker_cli_events_test.go:76: Container run with command blerg should have failed, but it did not
--- FAIL: TestPsListContainersSize (1.19s)
	docker_cli_ps_test.go:281: Expected size "2 B", got "80 B"
--- FAIL: TestPsListContainersFilterStatus (2.21s)
	docker_cli_ps_test.go:322: Expected id 7747caa688f6, got 7747caa688f6
		1e76d0fa0ff4 for exited filter, output: "7747caa688f6\n1e76d0fa0ff4\n"
--- FAIL: TestRunCreateVolumeWithSymlink (1.96s)
	docker_cli_run_test.go:534: [run] err: exit status 255, exitcode: 255
--- FAIL: TestRunVolumesFromSymlinkPath (2.42s)
	docker_cli_run_test.go:578: [run] (volume) err: exit status 255, exitcode: 255
--- FAIL: TestRunEnvironment (1.23s)
	docker_cli_run_test.go:762: Wrong environment: should be 8 variables, not: ", FALSE=true, HOME=/root, HOSTNAME=testing, PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin, TRICKY=tri, TRUE=false, cky, container=lxc"
--- FAIL: TestRunCapDropInvalid (1.29s)
	docker_cli_run_test.go:898: <nil> bin
		dev
		etc
		home
		lib
		lib64
		linuxrc
		media
		mnt
		opt
		proc
		root
		run
		sbin
		sys
		tmp
		usr
		var
		
--- FAIL: TestRunCapDropALLCannotMknod (1.01s)
	docker_cli_run_test.go:938: <nil> ok
		
--- FAIL: TestRunCapDropALLAddMknodCannotMknod (1.09s)
	docker_cli_run_test.go:953: exit status 1 2015/01/12 17:07:37 setup user setgroups operation not permitted
		lxc-start: lxc_start.c: main: 342 The container failed to start.
		lxc-start: lxc_start.c: main: 346 Additional information can be obtained by setting the --logfile and --logpriority options.
		
--- FAIL: TestRunCapAddInvalid (1.36s)
	docker_cli_run_test.go:970: <nil> bin
		dev
		etc
		home
		lib
		lib64
		linuxrc
		media
		mnt
		opt
		proc
		root
		run
		sbin
		sys
		tmp
		usr
		var
		
--- FAIL: TestRunCapAddALLCanDownInterface (1.04s)
	docker_cli_run_test.go:995: exit status 255 lxc-start: conf.c: dropcaps_except: 2507 unknown capability block_suspend
		lxc-start: conf.c: lxc_setup: 4244 failed to keep requested caps
		lxc-start: start.c: do_start: 688 failed to setup the container
		lxc-start: sync.c: __sync_wait: 51 invalid sequence number 1. expected 2
		lxc-start: start.c: __lxc_start: 1080 failed to spawn 'e928113e2c2ed77f034bfc725f640ed2a1027ba4f24406ee90670b35f612e195'
		lxc-start: lxc_start.c: main: 342 The container failed to start.
		lxc-start: lxc_start.c: main: 346 Additional information can be obtained by setting the --logfile and --logpriority options.
--- FAIL: TestRunSysNotWritableInNonPrivilegedContainers (1.01s)
	docker_cli_run_test.go:1056: sys should not be writable in a non privileged container
--- FAIL: TestRunProcNotWritableInNonPrivilegedContainers (1.04s)
	docker_cli_run_test.go:1078: proc should not be writable in a non privileged container
--- FAIL: TestRunWriteHostsFileAndNotCommit (1.18s)
	docker_cli_run_test.go:1857: diff should be empty
--- FAIL: TestRunWriteHostnameFileAndNotCommit (1.26s)
	docker_cli_run_test.go:1886: diff should be empty
--- FAIL: TestRunWriteResolvFileAndNotCommit (1.17s)
	docker_cli_run_test.go:1915: diff should be empty
--- FAIL: TestRunSetMacAddress (0.95s)
	docker_cli_run_test.go:2146: Set MAC address with --mac-address failed. The container has an incorrect MAC address: "8a:b5:b8:9d:02:27", expected: "12:34:56:78:9a:bc"
--- FAIL: TestRunUnknownCommand (1.23s)
	docker_cli_run_test.go:2631: Container should not have been able to start!
--- FAIL: TestRunModeIpcHost (0.96s)
	docker_cli_run_test.go:2663: IPC different with --ipc=host ipc:[4026532093] != ipc:[4026532211]
--- FAIL: TestRunModeIpcContainer (2.26s)
	docker_cli_run_test.go:2712: IPC different with --ipc=container:5bb82a37dc7cea65eae2914bee4ac87fa09fef42106e083da6feef3ed6139f5d ipc:[4026532211] != ipc:[4026532277]
--- FAIL: TestRunNonLocalMacAddress (1.19s)
	docker_cli_run_test.go:2854: Output should have contained "00:16:3E:08:00:50": eth0      Link encap:Ethernet  HWaddr 72:7D:B4:17:D3:6C  
		          inet addr:10.0.2.90  Bcast:10.0.255.255  Mask:255.255.0.0
		          UP BROADCAST RUNNING MULTICAST  MTU:9001  Metric:1
		          RX packets:3 errors:0 dropped:0 overruns:0 frame:0
		          TX packets:2 errors:0 dropped:0 overruns:0 carrier:0
		          collisions:0 txqueuelen:1000 
		          RX bytes:258 (258.0 B)  TX bytes:180 (180.0 B)
		
		lo        Link encap:Local Loopback  
		          inet addr:127.0.0.1  Mask:255.0.0.0
		          UP LOOPBACK RUNNING  MTU:65536  Metric:1
		          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
		          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
		          collisions:0 txqueuelen:0 
		          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
		
		, <nil>
--- FAIL: TestSaveDirectoryPermissions (2.80s)
	docker_cli_save_load_test.go:425: failed to find the layer with the right content listing
```

I am not including the exec tests because obviously they should fail, but is there anything above that should fail as well, like the Mac address ipc ones, etc

@jfrazelle After the pull-requests I have up get merged, we will have 3 tests failures:
--- FAIL: TestRunModePidHost (1.20s)
--- FAIL: TestRunContainerWithReadonlyRootfs (1.50s)
--- FAIL: TestSaveDirectoryPermissions (3.77s)

This is down from the 25+ test breaks we had when you opened the issue.
These two are new features that will need time to get into lxc(I'll try)
--- FAIL: TestRunModePidHost (1.20s)
--- FAIL: TestRunContainerWithReadonlyRootfs (1.50s) 
Which leaves 1 break:
--- FAIL: TestSaveDirectoryPermissions (3.77s)

I will try fixing this, but if not, I'd still request a jenkins/drone.io notification to any pull request that breaks any other tests. 
What do you think?

--- a/integration-cli/docker_cli_create_test.go
+++ b/integration-cli/docker_cli_create_test.go
@@ -2,11 +2,12 @@ package main
 
 import (
 	"encoding/json"
-	"github.com/docker/docker/nat"
 	"os"
 	"os/exec"
 	"testing"
 	"time"
+
+	"github.com/docker/docker/nat"
 )
 
 // Make sure we can create a simple container with some args
@@ -227,8 +228,8 @@ func TestCreateEchoStdout(t *testing.T) {
 }
 
 func TestCreateVolumesCreated(t *testing.T) {
-	defer deleteAllContainers()
 	testRequires(t, SameHostDaemon)
+	defer deleteAllContainers()
 
 	name := "test_create_volume"
 	if out, _, err := runCommandWithOutput(exec.Command(dockerBinary, "create", "--name", name, "-v", "/foo", "busybox")); err != nil {

--- a/integration-cli/docker_cli_links_test.go
+++ b/integration-cli/docker_cli_links_test.go
@@ -30,9 +30,8 @@ func TestLinksEtcHostsRegularFile(t *testing.T) {
 }
 
 func TestLinksEtcHostsContentMatch(t *testing.T) {
-	defer deleteAllContainers()
-
 	testRequires(t, SameHostDaemon)
+	defer deleteAllContainers()
 
 	runCmd := exec.Command(dockerBinary, "run", "--net=host", "busybox", "cat", "/etc/hosts")
 	out, _, _, err := runCommandWithStdoutStderr(runCmd)
@@ -124,9 +123,8 @@ func TestLinksPingLinkedContainersAfterRename(t *testing.T) {
 }
 
 func TestLinksIpTablesRulesWhenLinkAndUnlink(t *testing.T) {
-	defer deleteAllContainers()
-
 	testRequires(t, SameHostDaemon)
+	defer deleteAllContainers()
 
 	dockerCmd(t, "run", "-d", "--name", "child", "--publish", "8080:80", "busybox", "sleep", "10")
 	dockerCmd(t, "run", "-d", "--name", "parent", "--link", "child:http", "busybox", "sleep", "10")

--- a/integration-cli/docker_cli_nat_test.go
+++ b/integration-cli/docker_cli_nat_test.go
@@ -9,7 +9,7 @@ import (
 )
 
 func TestNetworkNat(t *testing.T) {
-	testRequires(t, SameHostDaemon)
+	testRequires(t, SameHostDaemon, NativeExecDriver)
 	defer deleteAllContainers()
 
 	iface, err := net.InterfaceByName("eth0")

--- a/integration-cli/docker_cli_rm_test.go
+++ b/integration-cli/docker_cli_rm_test.go
@@ -8,8 +8,8 @@ import (
 )
 
 func TestRmContainerWithRemovedVolume(t *testing.T) {
-	defer deleteAllContainers()
 	testRequires(t, SameHostDaemon)
+	defer deleteAllContainers()
 
 	cmd := exec.Command(dockerBinary, "run", "--name", "losemyvolumes", "-v", "/tmp/testing:/test", "busybox", "true")
 	if _, err := runCommand(cmd); err != nil {

--- a/integration-cli/docker_cli_run_test.go
+++ b/integration-cli/docker_cli_run_test.go
@@ -1348,8 +1348,8 @@ func TestRunAddingOptionalDevices(t *testing.T) {
 }
 
 func TestRunModeHostname(t *testing.T) {
-	defer deleteAllContainers()
 	testRequires(t, SameHostDaemon)
+	defer deleteAllContainers()
 
 	cmd := exec.Command(dockerBinary, "run", "-h=testhostname", "busybox", "cat", "/etc/hostname")
 	out, _, err := runCommandWithOutput(cmd)
@@ -2764,6 +2764,7 @@ func TestRunExposePort(t *testing.T) {
 }
 
 func TestRunUnknownCommand(t *testing.T) {
+	testRequires(t, NativeExecDriver)
 	defer deleteAllContainers()
 	runCmd := exec.Command(dockerBinary, "create", "busybox", "/bin/nada")
 	cID, _, _, err := runCommandWithStdoutStderr(runCmd)
@@ -2791,8 +2792,8 @@ func TestRunUnknownCommand(t *testing.T) {
 }
 
 func TestRunModeIpcHost(t *testing.T) {
-	defer deleteAllContainers()
 	testRequires(t, SameHostDaemon)
+	defer deleteAllContainers()
 
 	hostIpc, err := os.Readlink("/proc/1/ns/ipc")
 	if err != nil {
@@ -2901,8 +2902,8 @@ func TestContainerNetworkMode(t *testing.T) {
 }
 
 func TestRunModePidHost(t *testing.T) {
+	testRequires(t, NativeExecDriver, SameHostDaemon)
 	defer deleteAllContainers()
-	testRequires(t, SameHostDaemon)
 
 	hostPid, err := os.Readlink("/proc/1/ns/pid")
 	if err != nil {
@@ -3047,8 +3048,8 @@ func TestRunNonLocalMacAddress(t *testing.T) {
 
 func TestRunNetHost(t *testing.T) {
 	testRequires(t, SameHostDaemon)
-
 	defer deleteAllContainers()
+
 	hostNet, err := os.Readlink("/proc/1/ns/net")
 	if err != nil {
 		t.Fatal(err)
@@ -3157,7 +3158,9 @@ func TestRunContainerWithWritableRootfs(t *testing.T) {
 }
 
 func TestRunContainerWithReadonlyRootfs(t *testing.T) {
+	testRequires(t, NativeExecDriver)
 	defer deleteAllContainers()
+
 	out, err := exec.Command(dockerBinary, "run", "--read-only", "--rm", "busybox", "touch", "/file").CombinedOutput()
 	if err == nil {
 		t.Fatal("expected container to error on run with read only error")

--- a/integration-cli/requirements.go
+++ b/integration-cli/requirements.go
@@ -1,8 +1,11 @@
 package main
 
 import (
+	"encoding/json"
 	"fmt"
+	"log"
 	"os/exec"
+	"strings"
 	"testing"
 )
 
@@ -15,6 +18,8 @@ type TestRequirement struct {
 
 // List test requirements
 var (
+	daemonExecDriver string
+
 	SameHostDaemon = TestRequirement{
 		func() bool { return isLocalDaemon },
 		"Test requires docker daemon to runs on the same machine as CLI",
@@ -36,6 +41,30 @@ var (
 			return err == nil
 		},
 		fmt.Sprintf("Test requires an environment that can host %s in the same host", v2binary),
+	}
+	NativeExecDriver = TestRequirement{
+		func() bool {
+			if daemonExecDriver == "" {
+				// get daemon info
+				body, err := sockRequest("GET", "/info", nil)
+				if err != nil {
+					log.Fatalf("sockRequest failed for /info: %v", err)
+				}
+
+				type infoJSON struct {
+					ExecutionDriver string
+				}
+				var info infoJSON
+				if err = json.Unmarshal(body, &info); err != nil {
+					log.Fatalf("unable to unmarshal body: %v", err)
+				}
+
+				daemonExecDriver = info.ExecutionDriver
+			}
+
+			return strings.HasPrefix(daemonExecDriver, "native")
+		},
+		"Test requires the native (libcontainer) exec driver.",
 	}
 )
 

